pkg/ingester/ingester.go

```diff
@@ -3764,6 +3758,41 @@ func (i *Ingester) ShutdownHandler(w http.ResponseWriter, _ *http.Request) {
 	w.WriteHeader(http.StatusNoContent)
 }
 
+// startReadRequest tries to start a read request.
+// If it was successful, startReadRequest returns a function that should be
+// called to finish the started read request once the request is completed.
+// If it wasn't successful, the causing error is returned. In this case no
+// function is returned.
+func (i *Ingester) startReadRequest() (func(error), error) {
+	var (
+		start                        = time.Now()
+		err                          error
+		acquiredCircuitBreakerPermit bool
+	)
+	acquiredCircuitBreakerPermit, err = i.circuitBreaker.tryAcquirePermit()
+	if err != nil {
+		return nil, err
+	}
+	// In case of error we need to ensure that the acquired permit is returned.
+	defer func() {
+		i.circuitBreaker.finishReadRequest(time.Since(start), err)
+	}()
+
+	err = i.checkAvailableForRead()
+	if err != nil {
+		return nil, err
+	}
+	err = i.checkReadOverloaded()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I have updated the implementation and put something similar to what you proposed. The only difference is that we need to call `i.circuitBreaker.finishReadRequest(time.Since(start), err)` only if `acquiredCircuitBreakerPermit` is true

Reply:
Thanks.